src/cast.rs

```diff
@@ -0,0 +1,958 @@
+unsafe impl<T, U> TryCastRaw<RawCow<*const U, U>> for RawCow<*const T, T>
+where
+  T: NoUninit,
+  U: AnyBitPattern,
+{
+  fn try_cast_raw(self) -> Result<RawCow<*const U, U>, PodCastError> {
+    Ok(match self {
+      Self::Borrowed(x) => RawCow::Borrowed(x.try_cast_raw()?),
+      Self::Owned(x) => RawCow::Owned(x.reinterpret()),
+    })
+  }
+}
+#[cfg(feature = "extern_crate_alloc")]
+unsafe impl<T, U> TryCastRaw<(*const U, usize, usize)>
+  for (*const T, usize, usize)
+where
+  T: NoUninit,
+  U: AnyBitPattern,
+{
+  fn try_cast_raw(self) -> Result<(*const U, usize, usize), PodCastError> {
+    let len = cast_len::<T, U>(self.1)?;
+    let cap = cast_len::<T, U>(self.2)?;
+    Ok((self.0 as *const U, len, cap))
+  }
+}
+
+#[cfg(feature = "extern_crate_alloc")]
+unsafe impl<T, U>
+  TryCastRaw<RawCow<(*const U, usize), (*const U, usize, usize)>>
+  for RawCow<(*const T, usize), (*const T, usize, usize)>
+where
+  T: NoUninit,
+  U: AnyBitPattern,
+{
+  fn try_cast_raw(
+    self,
+  ) -> Result<RawCow<(*const U, usize), (*const U, usize, usize)>, PodCastError>
+  {
+    static_assert!(AssertSameAlign(T, U));
+    Ok(match self {
+      Self::Borrowed(x) => RawCow::Borrowed(x.try_cast_raw()?),
+      Self::Owned(x) => RawCow::Owned(x.try_cast_raw()?),
+    })
+  }
+}
+
+/// Checks any constraints the container requires when casting between types.
+pub trait AssertClassContraints<T, U> {
+  const ASSERT: () = ();
+}
+impl<T, U> AssertClassContraints<T, U> for RefT {}
+impl<T, U> AssertClassContraints<T, U> for PtrT {}
+impl<T, U> AssertClassContraints<T, U> for NonNullT {}
+impl<T, U> AssertClassContraints<T, U> for AtomicPtrT {}
+impl<'a, C, T, U> AssertClassContraints<T, U> for OptionT<C>
+where
+  C: AssertClassContraints<T, U>,
+{
+  const ASSERT: () = C::ASSERT;
+}
+impl<'a, C, T, U> AssertClassContraints<T, U> for PinT<C>
+where
+  C: AssertClassContraints<T, U>,
+{
+  const ASSERT: () = C::ASSERT;
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for BoxT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for RcT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for RcWeakT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for ArcT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for ArcWeakT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T, U> AssertClassContraints<T, U> for VecT {
+  const ASSERT: () = static_assert!(AssertSameAlign(T, U));
+}
+#[cfg(feature = "extern_crate_alloc")]
+impl<T: ToOwned, U: ToOwned> AssertClassContraints<T, U> for CowT {}
```

Comment:
Somwhere needs to assert/check that the source `T` and destination `U` of a casted `Cow<[T]>` -> `Cow<[U]>` have the same alignment, otherwise you can cast between `Cow::Owned`s and get undefined behavior:

```rs
// tests/unified_tests.rs
#[cfg(all(feature = "unified_cast", feature = "extern_crate_alloc"))]
#[test]
fn test_cow() {
  extern crate alloc;
  use alloc::{borrow::Cow, vec};
  let x: Cow<[u32]> = Cow::Borrowed(&[42]);
  // This correctly succeeds
  let y: Cow<[u8]> = x.reinterpret_inner();

  let x: Cow<[u32]> = vec![42].into();
  // This succeeds, but should fail since Vec<u32> cannot be cast to Vec<u8>.
  let y: Cow<[u8]> = x.reinterpret_inner(); 
}
```

```rs
$ cargo +nightly miri test
test test_cow ... error: Undefined Behavior: incorrect layout on deallocation: alloc71289 has size 4 and alignment 4, but gave size 4 and alignment 1
// ...
```